### DISCUSSION TITLE
Linting python modules written in C

### DISCUSSION THREAD

**Discussion by Luka-** [microsoft/pylance-release#1756] [Q&A]
Hello,

I'm developing a Blender add-on. Blender has built-in python modules (such as gpu or bgl) that are written in C. When trying to import those, pylance complains with Pylance(reportMissingImports).

Is there a way to deal wit h this/fix it? 
For example, pylint has: https://pylint.pycqa.org/en/latest/technical_reference/c_extensions.html

**Comment by jakebailey** [comment]
If the file can be imported by Python, then it should be shown as not missing in Pylance. I think it would make sense to file an issue with your repro and logs to show what the problem is.

We won't get info out of the compiled code, but we should be able to know that the import will succeed.

**Comment by erictraut** [comment]
Compiled modules typically have platform-specific file extensions. For example, Linux and Mac us ".so" and Windows uses ".pyd". Pylance looks for these binary files when it resolves an import. When you install the blender library, do you see a file with the name `<module>.<pyd|so>`?

**Comment by gvanrossum** [comment]
I'm just guessing here, but could it be that Blender embeds Python and makes these modules available as *builtin* modules? That would mean they cannot be imported when you execute the extension code in the Python executable -- they would only work when it's imported by the special embedded Python interpreter used by Blender to run extensions.

**Reply by erictraut** [reply]
@Luka- if Guido's theory is correct, then you should be able to work around it by simply disabling the `reportMissingImports` diagnostic check for your project. If you want to use static type checking in your code, you'll also probably want to find (or create) type stubs that describe the interface to the Blender modules.

**Reply by Luka-** [reply]
Thanks for the replies. 

I guess they are builtin modules, they have no __file__ attribute and when using inspect.getfile() on them (from inside Blender's built in python console), it throws a "module is a built-in module" error. 

I have built blender from source, and there's a python3.9 executable among blender build files, however when running that one I still cannot import built-in modules (I thought built-in meant that the python interpreter itself was extended to include the module, so that's weird). So, the way I see it, this could work in one of 2 ways:

1. During start-up of Blender, Blender loads the previously compiled built-in modules into the python. In that case, I should be able to find the .so files as @erictraut has suggested.

2. Perhaps more likely, the Blender binary itself contains the python interpreter together with the modules loaded in, which is not available outside blender. So, the python binary in the build files was really just a building artifact that was later infused to create the final executable (which is what @gvanrossum was meaning, I pressume, by "special embedded Python interpreter"). In this case, maybe I could find the built-in module c file, which during the building process should produce .so file to be linked, and then I copy it somewhere (or maybe I can just dig them out of build_files). I'm totally unfamiliar with the process here, so I'm just guessing, but please feel free to correct me if I'm wrong. 

I'll dig into this a bit more, but thank you for the answers, I have solutions if nothing else works. In either case, this isn't really a pylance issue anymore...

**Reply by Arch-Droid** [reply]
@Luka- Did you end up finding a solution for this? I was quite curious about it since it really helps to quickly understand the core modules of blender...

**Comment by gvanrossum** [comment]
It seems Blender is open source, so you should be able to find how it is built in its source code.
